runtime/lua/vim/hl.lua

```diff
@@ -108,38 +105,38 @@ function M.range(bufnr, ns, higroup, start, finish, opts)
     end
   end
 
-  if range_timer and not range_timer:is_closing() then
-    range_timer:close()
-    assert(range_hl_clear)
-    range_hl_clear()
-  end
-
-  range_hl_clear = function()
-    range_timer = nil
-    range_hl_clear = nil
-    pcall(vim.api.nvim_buf_clear_namespace, bufnr, ns, 0, -1)
-    pcall(vim.api.nvim__ns_set, { wins = {} })
-  end
-
+  local extmarks = {} --- @type integer[]
   for _, res in ipairs(region) do
     local start_row = res[1][2] - 1
     local start_col = res[1][3] - 1
     local end_row = res[2][2] - 1
     local end_col = res[2][3]
-    api.nvim_buf_set_extmark(bufnr, ns, start_row, start_col, {
-      hl_group = higroup,
-      end_row = end_row,
-      end_col = end_col,
-      priority = priority,
-      strict = false,
-    })
+    table.insert(
+      extmarks,
+      api.nvim_buf_set_extmark(bufnr, ns, start_row, start_col, {
+        hl_group = higroup,
+        end_row = end_row,
+        end_col = end_col,
+        priority = priority,
+        strict = false,
+      })
+    )
+  end
+
+  local range_hl_clear = function()
+    for _, mark in ipairs(extmarks) do
+      api.nvim_buf_del_extmark(bufnr, ns, mark)
+    end
   end
 
   if timeout ~= -1 then
-    range_timer = vim.defer_fn(range_hl_clear, timeout)
+    local range_timer = vim.defer_fn(range_hl_clear, timeout)
+    return range_timer, range_hl_clear
   end
 end
 
+local range_timer --- @type uv.uv_timer_t
+local range_hl_clear --- @type fun()?
 local yank_ns = api.nvim_create_namespace('nvim.hlyank')
```

Comment:
These now "belong" to `on_yank`, right? So maybe it makes sense to rename them with the same prefix?

```
local yank_timer
local yank_hl_clear
```